dlrover/trainer/torch/elastic.py

```diff
@@ -0,0 +1,270 @@
+        self.gradient_state = GradientState()
+        self.gradient_accumulation_steps = 1
+
+    def prepare(self, optimizer, lr_scheduler):
+        optimizer = _ElasticOptimizer(optimizer)
+        lr_scheduler = _ElasticOptimizer(optimizer)
+        self._set_gradient_accumulation_steps()
+        return optimizer, lr_scheduler
+
+    @contextmanager
+    def step(self, fix_total_batch_size=True):
+        """
+        A context manager that will lightly wrap around and to keep
+        the global batch size fixed when the number of worker changes.
+
+        Args:
+            fix_total_batch_size (`bool`): Whether to keep the total
+            batch size fixed when the optimizer steps. If True,
+            the context manager will perform gradient accumulation and
+            synchronize gradients when the accumulated batch size if
+            equal to the global batch size.
+
+        Example:
+        ```python
+        >>> from dlrover.trainer.torch.elastic import ElasticTrainer
+        >>> elastic_trainer = ElasticTrainer(model, dataloader)
+        >>> optimizer, scheduler = elastic_trainer.prepare(
+                optimizer, scheduler
+            )
+        >>> for input, output in dataloader:
+        ...     with elastic_trainer.step():
+        ...         outputs = model(input)
+        ...         loss = loss_func(outputs)
+        ...         loss.backward()
+        ...         optimizer.step()
+        ...         scheduler.step()
+        ...         optimizer.zero_grad()
+        ```
+        """
+        self._before_step(fix_total_batch_size)
+        context = contextlib.nullcontext
+        if self.gradient_state.sync_gradients:
+            context = getattr(self.model, "no_sync", context)
+
+        with context():
+            yield
+            self._after_step()
+
+    @property
+    def num_steps(self):
+        return self.gradient_state.num_steps
+
+    def _before_step(self, fix_total_batch_size):
+        """Sets the right `sync_gradients` and either resets
+        or increases `self.step`"""
+        self.gradient_state.num_backward_steps += 1
+        if not fix_total_batch_size:
+            self.gradient_state.sync_gradients = True
+        else:
+            self.gradient_state.check_sync_gradient(
+                self.gradient_accumulation_steps
+            )
+
+    def _after_step(self):
+        if isinstance(self._dataset, ElasticDataset):
+            self._dataset.report_batch_done()
+        if self.gradient_state.sync_gradients:
+            self.gradient_state.num_steps += 1
+
+    def _do_checkpoint(self):
+        """The job master of DLRover will notify the worker to do chekpoint
+        before it changes workers."""
+        pass
+
+    def _set_gradient_accumulation_steps(self):
+        max_worker_num = int(os.getenv("WORKER_NUM", 0))
+        if max_worker_num == 0:
+            self.gradient_accumulation_steps = 1
+
+        cur_world_size = int(os.getenv("WORLD_SIZE", 1))
```

Comment:
Who set WORLD_SIZE/RANK env?

Maybe better to use dist.get_world_size/get_rank.

Also, this assumes that data parallel is used. If other parallel method is used, this logic is incorrect.